sessionctx/variable/session.go

```diff
@@ -1907,3 +1923,85 @@ func (s *SessionVars) SlowLogFormat(logItems *SlowQueryLogItems) string {
 func writeSlowLogItem(buf *bytes.Buffer, key, value string) {
 	buf.WriteString(SlowLogRowPrefixStr + key + SlowLogSpaceMarkStr + value + "\n")
 }
+
+func (s *SessionVars) ResetTxnTimes() {
+	s.StartTimes = s.StartTimes[:0]
+	s.EndTimes = s.EndTimes[:0]
+}
+
+// ReportTxnAndLockDuration reports txn and the pessimistic locks' idle and duration to metric server.
+// Run this function in another goroutine to avoid it blocks the Commit or Rollback statements.
+func (s *SessionVars) ReportTxnAndLockDuration(startTS uint64, keysArr []kv.LockedKeys) {
+	if len(s.EndTimes) == 0 {
+		return
+	}
+	// copy the slice because it may be cleaned in other goroutines
+	startTimes := make([]time.Time, len(s.StartTimes))
+	copy(startTimes, s.StartTimes)
+	endTimes := make([]time.Time, len(s.EndTimes))
+	copy(endTimes, s.EndTimes)
+	// the rest works can be executed in another goroutine, so that it won't block the session goroutine.
+	go func() {
+		if len(startTimes) != len(endTimes)+1 {
+			panic(fmt.Sprintf("incorrect log durations, startTimes: %v, endTimes: %v", startTimes, endTimes))
+		}
+		var (
+			index = len(endTimes) - 1
+			// record idle duration since last statement's end to commit/rollback statement
+			idleDuration = startTimes[index+1].Sub(endTimes[index])
+			execDuration time.Duration
+		)
+		cfg := config.GetGlobalConfig()
+		// TODO: slow query is used to measure single statement, instead of long transaction, choose a better threshold.
+		threshold := time.Duration(atomic.LoadUint64(&cfg.Log.SlowThreshold)) * time.Millisecond
+		printSlowLog := startTimes[index+1].Sub(endTimes[0]) > threshold
+		var forUpdateTSBuilder strings.Builder
+		lockCnt := 0
+		for i := len(keysArr) - 1; i >= 0; i-- {
+			lockedKeys := keysArr[i]
```

Comment:
Would it be risky accessing the key reference concurrently if the `kv.transaction` is already recycled?